modules/core/arrow-test/src/main/kotlin/arrow/test/laws/SemialignLaws.kt

```diff
@@ -0,0 +1,170 @@
+
+    // toListOf (folded . here) (align x y)
+    val middle: List<A> = toList(align(a, b).map { it.justLeft() }).flattenOption()
+
+    // mapMaybe justHere (toList (align x y))
+    val right: List<A> = toList(align(a, b)).filterMap { it.justLeft() }
+
+    left == right && left == middle
+  }
+}
+
+// http://hackage.haskell.org/package/these-0.8/docs/src/Data.These.Combinators.html#justHere
+
+fun <A, B> Ior<A, B>.justLeft(): Option<A> =
+  fold({ Some(it) }, { None }, { a, _ -> Some(a) })
+
+// http://hackage.haskell.org/package/these-1.0.1/docs/src/Data.These.html#line-289
+
+fun <A, B, C> Ior<Ior<A, B>, C>.assoc(): Ior<A, Ior<B, C>> =
+  when (this) {
+    is Ior.Left -> when (val inner = this.value) {
+      is Ior.Left -> Ior.Left(inner.value)
+      is Ior.Right -> Ior.Right(Ior.Left(inner.value))
+      is Ior.Both -> Ior.Both(inner.leftValue, Ior.Left(inner.rightValue))
+    }
+    is Ior.Right -> Ior.Right(Ior.Right(this.value))
+    is Ior.Both -> when (val inner = this.leftValue) {
+      is Ior.Left -> Ior.Both(inner.value, Ior.Right(this.rightValue))
+      is Ior.Right -> Ior.Right(Ior.Both(inner.value, this.rightValue))
+      is Ior.Both -> Ior.Both(inner.leftValue, Ior.Both(inner.rightValue, this.rightValue))
+    }
+  }
+
+fun <A, B, C> Ior<A, Ior<B, C>>.unassoc(): Ior<Ior<A, B>, C> =
+  when (this) {
+    is Ior.Left -> Ior.Left(Ior.Left(this.value))
+    is Ior.Right -> when (val inner = this.value) {
+      is Ior.Left -> Ior.Left(Ior.Right(inner.value))
+      is Ior.Right -> Ior.Right(inner.value)
+      is Ior.Both -> Ior.Both(Ior.Right(inner.leftValue), inner.rightValue)
+    }
+    is Ior.Both -> when (val inner = this.rightValue) {
+      is Ior.Left -> Ior.Left(Ior.Both(this.leftValue, inner.value))
+      is Ior.Right -> Ior.Both(Ior.Left(this.leftValue), inner.value)
+      is Ior.Both -> Ior.Both(Ior.Both(this.leftValue, inner.leftValue), inner.rightValue)
+    }
+  }
+
+fun <A, B, C> Either<Either<A, B>, C>.assoc(): Either<A, Either<B, C>> =
+  when (this) {
+    is Either.Left -> when (val inner = this.a) {
+      is Either.Left -> Either.Left(inner.a)
+      is Either.Right -> Either.Right(Either.left(inner.b))
+    }
+    is Either.Right -> Either.Right(Either.Right(this.b))
+  }
+
+fun <A, B, C> Either<A, Either<B, C>>.unassoc(): Either<Either<A, B>, C> =
+  when (this) {
+    is Either.Left -> Either.Left(Either.Left(this.a))
+    is Either.Right -> when (val inner = this.b) {
+      is Either.Left -> Either.Left(Either.Right(inner.a))
+      is Either.Right -> Either.Right(inner.b)
+    }
+  }
+
+fun <A, B, C> Tuple2<Tuple2<A, B>, C>.assoc(): Tuple2<A, Tuple2<B, C>> =
+  this.a.a toT (this.a.b toT this.b)
+
+fun <A, B, C> Tuple2<A, Tuple2<B, C>>.unassoc(): Tuple2<Tuple2<A, B>, C> =
+  (this.a toT this.b.a) toT this.b.b
+
+fun <A, T> Const<Const<A, T>, T>.assoc(): Const<A, T> =
+  this.value()
+
+fun <A, T> Const<A, T>.unassoc(): Const<Const<A, T>, A> =
+  Const.just(this)
```

Comment:
Are those used anywhere? Maybe we can move them to some utility file or something.

Reply:
assoc/unassoc and justLeft functions in SemialignLaws. yes, it was my thought too to move them somewhere else. parts of them are not used (also did it as an exercise). would it be ok to put them into the respective extensions files?

Reply:
I don't think the either assoc/unassoc has any use in current tests, but the tuple one could definitely be used to implement the bijection method for any functor: `map(assoc)` should be equivalent. It is used a few times in some laws for collections (and probably others) and is basically duplicate code. It has to be moved to a utility file tho. If you can come up with a use case for use outside of tests we can also move this to the datatypes themselves.

Reply:
one of the assoc functions was used in the tests. i implemented the others as an exercise too without using them anywhere.

i removed the unused assoc/unassoc functions for now and changed the visibilty of the remaining two functions to private as they are only used from within the Laws file. is that acceptable for you?